app/containers/ChatContainer/saga.js

```diff
@@ -1,6 +1,29 @@
-// import { take, call, put, select } from 'redux-saga/effects';
+import { all, call, takeEvery, take, put } from 'redux-saga/effects';
+import { eventChannel } from 'redux-saga';
+import ws from 'services/websocket';
+import { SEND_MESSAGE_ACTION } from './constants';
+import { receiveMessageAction } from './actions';
 
-// Individual exports for testing
-export default function* defaultSaga() {
-  // See example in containers/HomePage/saga.js
+
+const sendMessage = ({ payload }) => {
+  ws.send(payload);
+};
+
+const watchMessages = eventChannel((emit) => {
+  ws.onmessage = (evt) => emit(evt);
+  return () => console.log('unsubscribe from ws');
```

Comment:
is `watchMessages` persistent for duration of the chat container? hence why we return an "unsubscribe" after the `eventChannel` and presumably the chat ends

Reply:
That's correct - this will get evaluated, and the channel created, when the code including this file is evaluated (currently from ChatContainer). The eventChannel will exist until we explicitly destroy it, and the API for destroying channels is to return a function that is invoked upon destruction.

We would need to call watchMessages.close() to invoke this behavior. We can also listen to the ws.onend (or onclosed? I forget) and emit END so that our channel can notify subscribers of completion.

See: https://github.com/redux-saga/redux-saga/blob/master/docs/advanced/Channels.md#using-the-eventchannel-factory-to-connect-to-external-events

Reply:
ok got it. we are not currently closing this `eventChannel` right now so will need to implement

Reply:
At some point, maybe. AFAIK we'd want to leave the socket open for the lifetime of the chat session. We should be multiplexing our socket so that we only have the single connection handling all communication. The only case where we'd want to disconnect would be on a session change - the user logs out (or maybe if they close all chat windows or something).

Not cleaning up our sockets will result in gradually increasing memory usage, but we shouldn't rely on the browser to always clean up connections (the `beforeunload` event can be finnicky and should not be relied on) - instead we should have the server send heartbeat messages to connected clients and clean up connections if they have been inactive for a certain amount of time. On the client side, we should always be testing for the presence of an active connection before sending a message.

Out of scope for here, though. That's for when we really hammer out our netcode.